dali/python/nvidia/dali/shared_batch.py

```diff
@@ -0,0 +1,262 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nvidia.dali import shared_mem, tensors
+from nvidia.dali import types
+import pickle
+
+
+np = None
+
+
+def _div_ceil(a, b):
+    """Calculate ceil of a/b without decaying to float."""
+    return -(-a // b)
+
+def _align_up(x, alignment):
+    """ Align x up to multiple of alignment"""
+    return _div_ceil(x, alignment) * alignment
+
+
+def import_numpy():
+    global np
+    if np is None:
+        try:
+            import numpy as np
+        except ImportError:
+            raise RuntimeError('Could not import numpy. Please make sure you have numpy '
+                               'installed before you use parallel mode.')
+
+
+class SampleMeta:
+    """Metadata describing serialized sample in a memory buffer.
+
+    It is passed through memory, stored after sample it describes."""
+
+    def __init__(self, offset, shape, dtype, nbytes):
+        self.shape = shape
+        self.dtype = dtype
+        self.offset = offset
+        self.nbytes = nbytes
+
+    @classmethod
+    def from_np(cls, offset, np_array):
+        return cls(offset, np_array.shape, np_array.dtype, np_array.nbytes)
+
+
+class SharedBatchMeta:
+    """Container for essential meta data about batch written into shared memory.
+    Contains id of the memory chunk, size of the buffer, and offset of more detailed
+    meta data passed through shared memory (SampleMeta).
+    """
+
+    def __init__(self, mem_chunk_id, capacity, meta_offset, meta_size):
+        self.mem_chunk_id = mem_chunk_id
+        self.capacity = capacity
+        self.meta_offset = meta_offset
+        self.meta_size = meta_size
+
+    @classmethod
+    def from_writer(cls, writer):
+        return cls(writer.mem_batch.mem_chunk_id, writer.mem_batch.capacity, writer.data_size,
+                   writer.meta_data_size)
+
+
+def deserialize_sample(buffer: shared_mem.SharedMem, sample):
+    if isinstance(sample, SampleMeta):
+        offset = sample.offset
+        assert offset % sample.dtype.itemsize == 0, "Sample offset is misaligned."
+        buffer = buffer.buf[offset:offset + sample.nbytes]
+        return np.ndarray(sample.shape, dtype=sample.dtype, buffer=buffer)
+    if isinstance(sample, (tuple, list,)):
+        return type(sample)(deserialize_sample(buffer, part) for part in sample)
+    return sample
+
+
+def deserialize_sample_meta(buffer: shared_mem.SharedMem, shared_batch_meta: SharedBatchMeta):
+    """Helper to deserialize SampleMeta from memory based on SharedBatchMeta.
+    """
+    sbm = shared_batch_meta
+    if sbm.meta_size == 0:
+        return []
+    pickled_meta = buffer.buf[sbm.meta_offset:sbm.meta_offset + sbm.meta_size]
```

Comment:
A certain layout is assumed. Can we have that abstracted away to one place? Or this should be the place?

Reply:
We do the serialize and deserialize in this file. The layout is to pack the buffer with the samples first and than their descriptors (and this is also reflected in the simpler metadata that contains the size and offset to this split). I can maybe do more, but it's still abstracted from the use.